src/services/webhdfs/backend.rs

```diff
@@ -0,0 +1,1026 @@
+    /// If both are set, delegation token will be used. And username
+    /// will be ignored.
+    pub fn user(&mut self, user: &str) -> &mut Self {
+        if user.is_empty() {
+            self.user = Some(user.to_string());
+        }
+        self
+    }
+
+    /// Set the proxy user of this backend
+    /// # Note
+    /// The builder prefers using delegation token,
+    /// If both are set, delegation token will be used. And do_as
+    /// will be ignored
+    pub fn doas(&mut self, doas: &str) -> &mut Self {
+        if !doas.is_empty() {
+            self.doas = Some(doas.to_string());
+        }
+        self
+    }
+
+    /// Set th CSRF prevention header name
+    /// refer to [CSRF prevention](https://hadoop.apache.org/docs/r3.3.3/hadoop-project-dist/hadoop-hdfs/WebHDFS.html#Cross-Site_Request_Forgery_Prevention) for detail
+    ///
+    /// # Note:
+    /// If not set, the default value is `X-XSRF-HEADER`
+    pub fn csrf(&mut self, csrf: &str) -> &mut Self {
+        if !csrf.is_empty() {
+            self.csrf = Some(csrf.to_string());
+        }
+        self
+    }
+
+    /// build the backend
+    ///
+    /// # Note:
+    /// when building backend, the built backend will check if the root directory
+    /// exits.
+    /// if the directory does not exits, the directory will be automatically created
+    /// if the root path is occupied by a file, a failure will be returned
+    pub fn build(&mut self) -> Result<impl Accessor> {
+        debug!("building backend: {:?}", self);
+
+        let root = normalize_root(&self.root.take().unwrap_or_default());
+        debug!("backend use root {}", root);
+
+        let endpoint = match self.endpoint.take() {
+            Some(e) => e,
+            None => WEBHDFS_DEFAULT_ENDPOINT.to_string(),
+        };
+        debug!("backend use endpoint {}", endpoint);
+
+        let csrf = match self.csrf.take() {
+            Some(c) => c,
+            None => "X-XSRF-HEADER".to_string(),
+        };
+
+        let auth = {
+            if let Some(token) = self.delegation.take() {
+                Some(WebHdfsAuth::new_delegation(&token))
+            } else if let Some(username) = self.user.take() {
+                let doas = if let Some(doas) = self.doas.take() {
+                    doas
+                } else {
+                    String::new()
+                };
+                Some(WebHdfsAuth::new_user(&username, &doas))
+            } else {
+                debug!("neither delegation token nor username is set");
+                None
+            }
+        };
+        let auth_str = auth.map(|s| s.auth_str());
+
+        let client = HttpClient::new()?;
+
+        let backend = apply_wrapper(Backend {
+            root: root.clone(),
+            endpoint,
+            auth: auth_str,
+            client,
+            csrf,
+        });
+
+        debug!("checking working directory: {}", root);
+        match backend.blocking_stat("/", OpStat::new()) {
```

Comment:
We don't have the plan to implement blocking API for HTTP-based services. Please consider wrap them inside `tokio::sync::OnceCell`.

Reply:
I removed the exposed blocking API and moved their functionalities to some helper functions, since checking and creating the root when building backends needs blocking HTTP transmissions.